File: CSharp/567-PermutationInString/567PermutationInString.cs
// Sliding Window
// Check the frequency of characters in s1 and the current window in s2 that is the size of s1.

// Time Complexity: O(n)
// n is the length of string s2. worst case doesn't contain permutation so you have to go through the whole thing.

// Space Complexity: O(1)
// We use the constant same 2 int arrays no matter the size of the input. 

public class Solution {
    public bool CheckInclusion(string s1, string s2) {
        if (s1.Length > s2.Length) return false;
        int[] s1Count = new int[26];
        int[] windowCount = new int[26];
        for (int i = 0; i < s1.Length; i++){
            s1Count[s1[i] - 'a']++;
            windowCount[s2[i] - 'a']++;
        } 
        bool AreArraysEqual(int[] arr1, int[] arr2){
            for (int i = 0; i < 26; i++){
                if (arr1[i] != arr2[i]) return false;
            }
            return true;
        }

        for (int i = 0; i < s2.Length - s1.Length; i++){
            if (AreArraysEqual(s1Count, windowCount)) return true;
            windowCount[s2[i + s1.Length] - 'a']++;
            windowCount[s2[i] - 'a']--;
        }
        return AreArraysEqual(s1Count, windowCount);
    }
}

File: Javascript/567-PermutationInString/567PermutationInString.js
/* Sliding Window
Time Complexity: O(n)
have go go through the string to count frequency
(a little unsure about explanation)

Space Complexity: O(1)
While you do need arrays to hold the counters, they are always size 26 no matter the size of the inputs.
*/

/**
 * @param {string} s1
 * @param {string} s2
 * @return {boolean}
 */
var checkInclusion = function(s1, s2) {
    if (s1.length > s2.length) {
        return false;
    }

    let s1Count = new Array(26).fill(0);
    let s2Count = new Array(26).fill(0);
    for (let i = 0; i < s1.length; i++) {
        s1Count[s1.charCodeAt(i) - 97]++;
        s2Count[s2.charCodeAt(i) - 97]++;
    }

    let matches = 0;
    for (let i = 0; i < 26; i++) 
        if (s1Count[i] === s2Count[i]) matches++;
        
    let l = 0;
    for (let r = s1.length; r < s2.length; r++) {
        if (matches === 26) return true;

        let index = s2.charCodeAt(r) - 97;
        s2Count[index]++;
        if (s1Count[index] === s2Count[index])
            matches++;
        else if (s1Count[index] + 1 === s2Count[index])
            matches--;

        index = s2.charCodeAt(l) - 97;
        s2Count[index]--;
        if (s1Count[index] === s2Count[index]) 
            matches++;
        else if (s1Count[index] - 1 === s2Count[index])
            matches--;
        l++;
    }
    return matches === 26;
};